---
title: "MulEA"
author: "Wiktor Jurkowski, Eszter Ari, David Fazekas, Leila Gul, Marton Oelbei, Cezary Turek"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{MulEA}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r global_options, include=TRUE, echo=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = TRUE, message = TRUE, error = FALSE)
```

   Functional interpretation of the biological data typically involves identifying key genes, molecules, reactions or pathways by finding non-random changes between two or more conditions or phenotypes. Performing enrichment analysis on set of molecules selected from  differential omics analysis is a method of choice. Among many packages that can be applied for this task, only few could be applied either to multiple species, ontology types or providing an access to multiple statistics.
   
   MulEA is addressing this gap in addition providing improved way to calculate correction for multiple testing that assume partial dependence between ontology terms and in result limits number of correct associations falsely scored as insignificant. Besides the commonly applied tests, MulEA provides a unique permutation based, empirical false discovery rate correction of the p-values to substitute the too conservative Bonferroni and Benjamini-Hochberg procedures.

   MulEA allows enrichment analysis using most popular gene and pathway ontologies (GO, KEGG, Reactome). In addition, one can test enrichment in genomic locations and in gene expression, protein domain, miRNA and transcription factors data bases, all created from publicly available resources and presented in standardized manner. Beyond genes or proteins, MulEA even allows working with basically any kind of data types, i.e. small molecules, chromosome region, enhancers, molecular interactions or any other information defined by the user.

   To analyse the data MulEA provide multiple types of statistics in one tool, which allows the user to calculate over-representations using the hypergeometric test, and enrichment analyses of ranked input by the Kolmogorov-Smirnov test.

   To conclude, MulEA is a comprehensive enrichment software that allows expansive analyses using diverse ontologies, statistical models and p-value correction procedures that can extend our understanding of the results of various high-throughput analyses and therefore expand our knowledge.
    
Implemented features include:

- [Mulea Input Output Data Workflow](#mulea-input-output-data-workflow)
    - [File](#file)
    - [Data Frame](#data-frame)
- [Data Analysis](#data-analysis)
    - [Set Based Tests](#set-based-tests)
        - [Hypergeometric Test](#hypergeometric-test)
        - [Multiple Comparisons Problem](#multiple-comparisons-problem)
    - [Ranked Based Tests](#ranked-based-tests)
        - [Kolmogorov Smirnov Test](#kolmogorov-smirnov-test)
        - [Subramanian Test](#subramanian-test)
    - [Graph Base](#graph-base)
- [Other](#other)




## Mulea Input Output Data Workflow
There are two ways to work with data in MulEA; directly from GMT files or with properly formated data frames. Detailed instructions are in paragraphs below. Remember that before starting work with MulEA you have to load and attach a package:
```{r, results = 'asis'} 
library(package="MulEA")
```


### File
The file which includes the ontology must be in GMT: Gene Matrix Transposed file format (*.gmt) ([format explanation](http://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#GMT:_Gene_Matrix_Transposed_file_format_.28.2A.gmt.29)). Example of the file is under MulEA installation directory in the extdata/ folder. To find the installation directory use `find.package("MulEA")` command.

In order to read the GMT file as a dataframe use: `MulEA::readGmtFileAsDataFrame()`. Method requires(*) one argument `gmtFilePath`:

- `gmtFilePath` - This is the path to the file, Example: `"R/MulEA/extdata/model.gmt"`

The ontology is then read into the data frame, such:
```{r, results = 'asis'} 
pathToModelGmtFile <- system.file(package="MulEA", "extdata", "model.gmt")
modelDfFromFile <- MulEA::readGmtFileAsDataFrame(gmtFilePath = pathToModelGmtFile)
```
```{r, results = 'asis', echo = FALSE} 
knitr::kable(modelDfFromFile, caption = "Model Data Frame")
```

If you would like to save the data frame that represents the ontology of the GMT file, use: `MulEA::saveDataFrameAsGmtFile()`. You have to provide two arguments `modelDF`, `gmtFilePath`.

- `modelDF` - ontology data frame which represents the GMT file.
- `gmtFilePath` - Path to file where you want to save the ontology. Example: `"R/MulEA/extdata/savedModel.gmt"`

```{r, eval=FALSE} 
   MulEA::saveDataFrameAsGmtFile(modelDF = modelDfFromFile, gmtFilePath = pathToModelGmtFile)  
```


### Data Frame
Data frame used in previous paragraph is the properly formatted data frame. You should use data frame in that format to be sure that results are proper and counting goes without any problems. Below is presented the structure of an data frame.

```{r, results = 'markup', echo=TRUE} 
modelDfFromFile <- MulEA::readGmtFileAsDataFrame(gmtFilePath = system.file(package="MulEA", "extdata", "model.gmt"))
str(modelDfFromFile)
```




## Data Analysis
"Analysis of data is a process of inspecting, cleansing, transforming, and modeling data with the goal of discovering useful information, suggesting conclusions, and supporting decision-making.", [Wikipedia](https://en.wikipedia.org/wiki/Data_analysis). MulEA performs data analysis on data frames. We implemented three types of tests base on contingency tables, rankings and topology of data.



### Set Based Tests
Hypothesis tests may be performed on data divided into sets. MulEA provides tests which are based on that sets of data. For now MulEA uses Hypergeometric test under `SetBasedTest` class. This class is highly configurable and open for us a possibility to adjust test's results by typical methods and newly presented permutation test.  
Before we perform this tests, we have to prepare data in proper form. Below is full data input prepered on our example data. Example data are small, to show results for user in a fast way.  

```{r, results = 'asis', echo = TRUE} 
modelDfFromFile <- MulEA::readGmtFileAsDataFrame(gmtFilePath = system.file(package="MulEA", "extdata", "model.gmt"))
dataFromExperiment <- c("FBgn0004407", "FBgn0010438", "FBgn0003742", "FBgn0029709", "FBgn0030341", "FBgn0037044", "FBgn0002887", "FBgn0028434", "FBgn0030170", "FBgn0263831")
dataFromExperimentPool <- unique(c(c("FBgn0033690", "FBgn0261618", "FBgn0004407", "FBgn0010438", "FBgn0032154", "FBgn0039930", "FBgn0040268", "FBgn0013674",
                                   "FBgn0037008", "FBgn0003116", "FBgn0037743", "FBgn0035401", "FBgn0037044", "FBgn0051005", "FBgn0026737", "FBgn0026751",
                                   "FBgn0038704", "FBgn0002887", "FBgn0028434", "FBgn0030170", "FBgn0263831", "FBgn0000579"),
                                 c("FBgn0066666", "FBgn0000000", "FBgn0099999", "FBgn0011111", "FBgn0022222", "FBgn0777777", "FBgn0333333", "FBgn0003742",
                                   "FBgn0029709", "FBgn0030341")))
```

#### Hypergeometric Test  

Basic usage of `SetBasedTest` class lokk like that:

```{r, results = 'asis', echo = TRUE} 
setBasedTest <- SetBasedTest(gmt = modelDfFromFile, testData = dataFromExperiment)
setBasedTestRes <- MulEA::runTest(setBasedTest)
```
In tests based on sets, Mulea will inform the user when tested data are outside of the model data by message presented above.
```{r, results = 'asis', echo = FALSE} 
knitr::kable(setBasedTestRes, caption = "Set Based Test Result Data Frame")
```

Usage with definition of pool data, which can be different that in presented model is presented below:

```{r, results = 'asis', echo = TRUE} 
setBasedTestWithPool <- SetBasedTest(gmt = modelDfFromFile, testData = dataFromExperiment, pool = dataFromExperimentPool)
setBasedTestWithPoolRes <- MulEA::runTest(setBasedTestWithPool)
```
```{r, results = 'asis', echo = FALSE} 
knitr::kable(setBasedTestWithPoolRes, caption = "Set Based Test Result Data Frame")
```

#### Multiple Comparisons Problem  

According to article [Multiple Comparisons Problem](https://en.wikipedia.org/wiki/Multiple_comparisons_problem) you can adjust p-values of tested ontologys by adding the `adjustMethod` argument to `SetBasedTest` class. The value of this argument can newly presented by MulEA method to adjust p-values besed on permutation test. To run that method, please use `"EszterPermutationTest"` as `adjustMethod` argument.
You can use common R approach and you can use one of the listed values: "holm", "hochberg", "hommel", "bonferroni", "BH", "BY", "fdr". References to this methods can be found here: [link](http://stat.ethz.ch/R-manual/R-devel/library/stats/html/p.adjust.html).

```{r, results = 'asis', echo = TRUE} 
setBasedTestWithPoolAndAdjust <- SetBasedTest(gmt = modelDfFromFile, testData = dataFromExperiment, pool = dataFromExperimentPool, adjustMethod = "BH")
setBasedTestWithPoolAndAdjustRes <- MulEA::runTest(setBasedTestWithPoolAndAdjust)
```
```{r, results = 'asis', echo = FALSE} 
knitr::kable(setBasedTestWithPoolAndAdjustRes, caption = "Set Based Test Result Data Frame")
```

Data frames with adjusted p-values contain one extra column, which include q-values.

`SetBasedTest` class constructor accepts list of arguments, such:

- `gmt` - required* - It is data.frame, which represents model. Read it from file or load from DBs.
- `testData` - required* - Vector of your experimental data. Example: `dataFromExperiment <- c("FBgn0004407", "FBgn0010438", "FBgn0003742")`.
- `pool = character()` - default: character() - It is vector of background - pool data to experiment data. Example: `dataFromExperimentPool <- c("FBgn0004407", "FBgn0010438", "FBgn0003742", "FBgn0003444", "FBgn0003333"")`
- `adjustMethod = NA` - default: NA - You can specify an algorithm which helps you with [Multiple Comparisons Problem](#multiple-comparisons-problem).

Column names presented in results data.frame are:

- `ontologyId` - input copy - Column copies from the input data frame. It include onlology ids. Could be for example ids from GO.
- `ontologyName` - input copy - Column copies from the input data frame. It include onlology names. Could be for example name from GO as "mitochondrion inheritance".
- `listOfValues` - input copy - Column copies from the input data frame. It include all symbols undet presented ontology id. Example from GO: FBgn0004407, FBgn0010438.
- `overlappingData` - output - Column includes set of intersection of list of values from model and provided by user experiment vector.
- `contingencyTable` - output - This colum presents contingeny tables used to count test.
- `p.value` - output - Cells of this column include counted p-value for provided data.
- `q.value` - output - Cells of this column include adjusted p-value according to the model. The result of adjustment is q-value.



### Ranked Based Tests
Fundament under ranked based tests is ranking. Ranking can be an ordered vector or any vector with vector of scores, both of them have to be the same length. For now MulEA provides you two ranked based tests. them are the Kolmogorov-Smirnov test and the Subramanian test. Both of them are encloused in `RankedBasedTest` class, which provides you a possibility to set input data and configur other parameters including used statistic method.

As privously, before running any tests you have to prepare proper input data. An example is presented below:

```{r, results = 'asis', echo = TRUE} 
modelDfFromFile <- MulEA::readGmtFileAsDataFrame(gmtFilePath = system.file(package="MulEA", "extdata", "model.gmt"))
dataFromExperiment <- c("FBgn0004407", "FBgn0010438", "FBgn0003742", "FBgn0029709", "FBgn0030341", "FBgn0037044", "FBgn0002887", "FBgn0028434", "FBgn0030170", "FBgn0263831")
dataFromExperimentScores <- c(0.09, 0.11, 0.15, 0.20, 0.21, 0.24, 0.28, 0.30, 0.45, 0.50)
```

#### Kolmogorov Smirnov Test  

Kolmogorov-Smirnov test is achieved by setting `method` argument to `"KS"`. It is also required to provide `testData` argument.

```{r, results = 'asis', echo=TRUE, warning=FALSE} 
rankedBasedTestKs <- RankedBasedTest(method = "KS", gmt = modelDfFromFile, testData = dataFromExperiment)
rankedBasedTestKsRes <- MulEA::runTest(rankedBasedTestKs)
```
During the execution of this chunk you will see many warnings from ks.test function from stats package. Warnings look like that:
```{r, results = 'asis', echo=TRUE} 
## Warning in ks.test(matchedFromModelDist, randomFromExperimentDist): cannot
## compute exact p-value with ties
```
The reason is that data which we are using in vignette are artificial one. They are constructed to show you how MulEA is working with understanding.
When you see any error or warnings when you are working with real data, it should be investigated. MulEA is not stopping warnings and error propagation from packages which are used. The interpretation of those messages belongs to user.


#### Subramanian Test

Subramanian method required `method` argument to be set to `"Subramanian"`. It;s also required `testData` and `scores` to be set. It is important that those two vectors have to be of the same lenght.

```{r, results = 'asis', echo=TRUE} 
rankedBasedTestSubramanian <- RankedBasedTest(method = "Subramanian", gmt = modelDfFromFile, testData = dataFromExperiment, scores = dataFromExperimentScores)
rankedBasedTestSubramanianRes <- MulEA::runTest(rankedBasedTestSubramanian)
```

Below list is the list of arguments accepted by the constructor of the `RankedBasedTest`:

- `method` - required* - It allows user to choose method, which will be used to count probabilities. For no it have to be one of this values "KS", "Subramanian".
- `gmt` - required* - It is data.frame, which represents model. Read it from file or load from DBs.
- `testData` - required* - Vector of your experimental data. Example: `dataFromExperiment <- c("FBgn0004407", "FBgn0010438", "FBgn0003742")`. In case of KS test it makes a role of ranking also. In Subramanian it creates a ranking with scores argument.
- `scores` - if method="Subramanian": required*, default: numeric() - This agument is vector of numbers. It creates a ranking with `testData` argument for Subramanian approach.
- `numberOfPermutations` - default: 1000 - This set number of permutations used to count p-value. You can speed up process of counting by set it to small value, remember that it can result with less accuracy.

Returned data frame from any ranked based test look like that (column specification):

- `ontologyId` - input copy - Column copies from the input data frame. It include onlology ids. Could be for example ids from GO.
- `ontologyName` - input copy - Column copies from the input data frame. It include onlology names. Could be for example name from GO as "mitochondrion inheritance".
- `listOfValues` - input copy - Column copies from the input data frame. It include all symbols undet presented ontology id. Example from GO: FBgn0004407, FBgn0010438.
- `p.value` - output - Cells of this column include counted p-value for provided data.

```{r, results = 'asis', echo = FALSE} 
knitr::kable(rankedBasedTestKsRes, caption = "Ranked Based Test Result Data Frame")
```
  
  

### Graph Based Test
(not implemented yet)




## Other
(not implemented yet)

